contracts/EntryPoint.sol

```diff
@@ -79,16 +81,56 @@ contract EntryPoint is StakeManager {
         require(success);
     }
 
+    struct AggregatorInfo {
+        IAggregator aggregator;
+        uint count;
+        bytes signature;
+    }
+
+    //run all aggregators, each on its respective wallets
+    function validateAggregatedSignatures(
+        UserOperation[] calldata ops,
+        AggregatorInfo[] calldata aggregators
+    ) private view returns (uint totalCount) {
+        if (aggregators.length == 0) return 0;
+        uint startPtr = getMemoryPointer();
+        AggregatorInfo[] memory _aggregators = aggregators;
+
+        uint index = 0;
+        uint loopPtr = getMemoryPointer();
+        for (uint i = 0; i < _aggregators.length; i++) {
+            restoreMemoryPointer(loopPtr);
+            AggregatorInfo memory agg = _aggregators[i];
+            uint count = agg.count;
+            agg.aggregator.validateSignatures(
+                copyOpsRange(agg.aggregator, ops, index, index + count),
+                agg.signature
+            );
+            index += count;
+        }
+        restoreMemoryPointer(startPtr);
+        return index;
+    }
+
     /**
      * Execute a batch of UserOperation.
      * @param ops the operations to execute
      * @param beneficiary the address to receive the fees
+     * @param aggregators list of aggregators to validate signatures
+     * @param execOrder execution order of userOps.
      */
-    function handleOps(UserOperation[] calldata ops, address payable beneficiary) public {
+    function handleOps(
+        UserOperation[] calldata ops,
+        address payable beneficiary,
+        AggregatorInfo[] calldata aggregators,
+        uint[] memory execOrder
+    ) public {
 
         uint256 opslen = ops.length;
         UserOpInfo[] memory opInfos = new UserOpInfo[](opslen);
 
+        uint totalAggregated = validateAggregatedSignatures(ops, aggregators);
```

Comment:
Another issue with not measuring gas at this phase is that if/when we make aggregators staked and permissionless, someone could write a "cost-saving" aggregator.  This aggregator would just call a validation function in the wallet, letting the wallet perform expensive validation at the bundler's expense, and then in the "official" validation it'll just pay and set the nonce.  It means that we may be stuck with a whitelist since someone will surely deploy such aggregator to perform validation at the bundler's expense.

With a whitelist this attack is not possible, but a wallet could still increase the cost to the bundler.  The wallet would return an aggregator and a BLS key, but the "config" part of the sig will be its usual ECDSA sig.  The bundler would spend eth on this redundant BLS validation which the wallet doesn't pay for.

I don't see an easy way to attribute aggregated validation to individual wallets without trusting the aggregator code (which wouldn't be great).  But not being able to attribute this gas, seems to cause some issues.

Reply:
1. validation is view function. you can waste gas, but not execute anything
2. aggregators are "whitelisted", in a sense that a aggregator's address is tied to the off-chain signature  code in the bundler .

Reply:
Yes, I meant wasting gas, not executing stuff.  Just griefing.  Since the aggregator has to call the wallet (e.g. to get the pubkey), it can't know the amount of gas the wallet will waste.  So maybe aggregators should put a gas limit on calls it makes to wallets.  And then they can advertise the max cost, and it can be added to preVerificationGas.

Re whitelist, in the future it doesn't have to be.  The aggregator could just be staked, and not have an off-chain signature code in the bundler, but offer a pure function in the aggregator contract, which does the off-chain work in the bundler.  Although in most cases it would make sense to optimize by implementing stuff in the bundler itself.